### ISSUE TITLE
Key error when using table alias

### ISSUE THREAD

**Issue by KiraThiessat** [rbergm/PostBOUND#2]
Using a table alias produces a KeyError:

<details>

![PB_Key_Error](https://user-images.githubusercontent.com/22939771/196508596-922ef90f-63e1-42a9-a1d5-37fa97054cb3.jpg)

Query 0a1a....:
![PB_Key_Error_2](https://user-images.githubusercontent.com/22939771/196509193-2be14785-774f-4683-930e-11378fe40590.jpg)

</details>

**Comment by rbergm** [comment]
Thank you for reporting, this should definitely not happen. Can you share your UES invocation? Are you using the `ues-generator.py` utility?

**Comment by KiraThiessat** [comment]
Sure:
`python ues/ues-generator.py --pattern "*.sql" --generate-labels --table-estimation precise --out stack-explicit.csv workloads/Stack-Queries/`

my workload queries are in the Stack-Queries folder which I added manually. Yes, I am using the `ues-generator.py` directly.

**Comment by rbergm** [comment]
Alright, thank you. This is definitively a bug in PostBOUND. I will take a look!

**Comment by rbergm** [comment]
I think I located the root cause of your reported issues and submitted a fix in 7a52fee14791fda5a993965fac1a0212c2591080. I would appreciate, if you could re-run your tests on actual instance of the Stack benchmark and report whether the issues persist (and which queries are affected).

**Comment by KiraThiessat** [comment]
The issue still persists as of the latest update:

<details>

![grafik](https://user-images.githubusercontent.com/22939771/196765423-e92f5765-42f5-4d2f-a555-017903576a61.png)

</details>

For your convenience, here is the link to the DB and query files: [Stack](https://rmarcus.info/stack.html)

**Comment by rbergm** [comment]
Ok, that's weird. Based on the output from `ues-generator.py` it is quite hard to figure out what is going wrong. On my machine, I am able to optimize query 0a1a... without any issues.

Could you maybe run the unit tests for the stack benchmark and tell me the error message you are getting. You can execute the tests from within the `ues/tests` directory like so: `python3 -m unittest test_ues.StackWorkloadOptimizationTests`

EDIT: btw, make sure to put a copy of your `.pyscopg_connection` in the `tests/` folder as well, the file should be called `.psycopg_connection_stack` in this case (other tests use different benchmarks in this folder, hence multiple connection files can exist)

EDIT 2: if you add `--verbose` to the test invocation, a bunch of additional information will be printed. This should be helpful as well

**Comment by KiraThiessat** [comment]
Ok, some more information on this topic. Apparently, my git didn't pull all recent changes. Now it did and some queries (like 0a1a) run. However, there is - of cource - another issue:

<details>

![Screenshot 2022-10-20 102152](https://user-images.githubusercontent.com/22939771/196900676-616459a6-65c9-493d-b6b3-36d9a8d8d63d.jpg)

</details>

So I tried running the test that I also pulled in the most recent update. Doing so resulted in this error:

<details>

![Screenshot 2022-10-20 103855](https://user-images.githubusercontent.com/22939771/196900722-6db991b6-ef09-45ac-b5b5-70bcccbe2e75.jpg)

</details>

It seems like the code is expecting another `.psycopg_connection` file somewhere. I have one in my `PostBOUND/` base directory as well as a `.psycopg_connection_stack` file in `PostBOUND/ues/tests/`

**Comment by rbergm** [comment]
Alright, that's at least some good news. Let's digest bit by bit:

>Now it did and some queries (like 0a1a) run. However, there is - of cource - another issue:

I will take a look at these issues -- although I have never seen the parts of the optimizer responsible for such error messages complain before. Seems like the Stack queries are a lot different from Join Order Benchmark and Star Schema Benchmark.

> It seems like the code is expecting another .psycopg_connection file somewhere. I have one in my PostBOUND/ base directory as well as a .psycopg_connection_stack file in PostBOUND/ues/tests/

Alright so this part is interesting: it is both a bug as well as poor wording on my end. I already fixed the bug and pushed an update to this repo. Now, concerning the poor wording:

**TL;DR:** also copy the `.psycopg_connection` file to the tests directory (in addition to the `*_stack` file)

The `.psycopg_connection` files have to be put in *all* directories where you want to execute *any* PostBOUND script from. I.e., if you run some code from the `ues/` directory, there has to be a connect file there. If you run some code in the root directory, there has to be a connect file there. This also applies to the `ues/tests/` directory. But in this case, you need the `.psycopg_connection_stack` file *in addition*. As long as your only using one database, all of these files can share the exact same content, they can even be symlinks to one another.

Why is this plethora of connect files necessary? A lot of PostBOUND code needs a working connection to the database to retrieve metadata such as index structures or table statistics. To prevent this db connection from being carried everywhere, a lot of functions use a default parameter for "easier" testing. If you look closely, you will often notice the pattern `dbs = db.DBSchema.get_instance()` in many function signatures. This DB schema is implemented as a [singleton pattern](https://en.wikipedia.org/wiki/Singleton_pattern) to ensure all instances of the code share the same database connection. Now, the implementation of `DBSchema` needs to connect to the database. This is achieved by reading the `.psycopg_connection` file per default. Since users of the DBSchema can be located in many different places, this file has to be located in the current working directory (which coincides with the path from which execution was started). Therefore, you need one file in the root directory, and another one in the tests directory.

However, this structure has one caveat: what happens, when one script tries to connect to multiple databases? For example, this is the case for the unit tests which evaluate the query optimization on different workloads: they need connections to an IMDB database for the Join Order Benchmark, or an TPCH database for the Star Schema Benchmark.  Therefore, multiple different connection files have to be put in the test directory and each unit test will overwrite the database connection based on his requirements. Hence, a `.pyscopg_connection_stack` file is needed in addition to the default file used by Python when the function parameters are evaluated.

One final remark: since its initial design, PostBOUND has become a lot more complex. The current DBSchema implementation and usage has outlived its usefulness and will be much improved in the future.

**Comment by rbergm** [comment]
Closing since this won't be an issue in the revamped PostBOUND implementation
